pyccl/halos/hbias.py

```diff
@@ -0,0 +1,289 @@
+                                 " rho = " + mass_def.rho_type)
+            self.mdef = mass_def
+        else:
+            self._default_mdef()
+        self._setup(cosmo)
+
+    def _default_mdef(self):
+        """ Assigns a default mass definition for this object if
+        none is passed at initialization.
+        """
+        self.mdef = MassDef('fof', 'matter')
+
+    def _setup(self, cosmo):
+        """ Use this function to initialize any internal attributes
+        of this object. This function is called at the very end of the
+        constructor call.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+        """
+        pass
+
+    def _check_mdef(self, mdef):
+        """ Return False if the input mass definition agrees with
+        the definitions for which this parametrization
+        works. True otherwise. This function gets called at the
+        start of the constructor call.
+
+        Args:
+            mdef (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            bool: True if the mass definition is not compatible with
+                this parametrization. False otherwise.
+        """
+        return False
+
+    def _get_consistent_mass(self, cosmo, M, a, mdef_other):
+        """ Transform a halo mass with a given mass definition into
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return np.log10(M_use)
+
+    def get_halo_bias(self, cosmo, M, a, mdef_other=None):
+        """ Returns the halo bias for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: halo bias.
+        """
+        M_use = np.atleast_1d(M)
+        logM = self._get_consistent_mass(cosmo, M_use,
+                                         a, mdef_other)
+
+        # sigma(M)
+        status = 0
+        sigM, status = lib.sigM_vec(cosmo.cosmo, a, logM,
+                                    len(logM), status)
+        check(status)
+
+        b = self.get_bsigma(cosmo, sigM, a)
+        if np.ndim(M) == 0:
+            b = b[0]
+        return b
+
+    def get_bsigma(self, cosmo, sigM, a):
```

Comment:
This routine and the one above it have force a specific functional form of the halo bias into the API. Do we want to be able to generalize to things like emulators later?

Reply:
Yeah, I thought about this. Not sure, but I think in that case those potential subclasses should just overload the `get_halo_bias` method.

Reply:
Would those sub classes be expect to implement `get_bsigma` even if they had overloaded `get_halo_bias`? If not, then `get_bsigma` should be named `_get_bsigma` so it is not part of the API.

Reply:
Good point. Done